### PR TITLE
MAID-803_Implement_Mpid_Manager_Vault_V2

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

File: src/maidsafe/vault/mpid_manager/database.cc [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+#include "maidsafe/common/utils.h"
+
+namespace maidsafe {
+
+namespace vault {
+
+MpidManagerDatabase::MpidManagerDatabase() : container_(), mutex_() {}
+
+void MpidManagerDatabase::Put(const MessageKey& key,
+                              const uint32_t size,
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/database.cc]: No need for `const` on `size`.

File: src/maidsafe/vault/mpid_manager/database.cc [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                              const GroupName& group_name) {
+  std::unique_lock<std::mutex> lock(mutex_);
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/database.cc]: Can't this just be a `lock_guard`?

File: src/maidsafe/vault/mpid_manager/database.h [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+using EntryByMpid = boost::multi_index::index<DatabaseEntrySet, EntryMpid_Tag>::type;
+using EntryByMpidIterator = DatabaseEntrySet::index<EntryMpid_Tag>::type::iterator;
+
+class MpidManagerDatabase {
+ public:
+  MpidManagerDatabase();
+
+  void Put(const MessageKey& key, const uint32_t size, const GroupName& mpid);
+  void Delete(const MessageKey& key);
+  bool Has(const MessageKey& key);
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/database.h:80]: Can't `Has` and the subsequent four functions all be `const`?

File: src/maidsafe/vault/mpid_manager/database.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+namespace vault {
+
+using GroupName = Identity;
+using MessageKey = Identity;
+using GKPair = std::pair<GroupName, MessageKey>;
+using DbTransferInfo = std::map<Identity, std::vector<GKPair>>;
+
+struct DatabaseEntry {
+  DatabaseEntry(const MessageKey& key_in,
+                const uint32_t size_in,
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/database.h]: No need for `const`.

File: src/maidsafe/vault/mpid_manager/handler.cc [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+#include "maidsafe/vault/mpid_manager/handler.h"
+
+#include "maidsafe/common/convert.h"
+
+namespace maidsafe {
+
+namespace vault {
+
+MpidManagerHandler::MpidManagerHandler(const boost::filesystem::path vault_root_dir,
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/handler.cc]: Should be `const boost::filesystem::path&` ?

**Comment by Fraser999** [comment]
I think there are a lot of places where we're using `Identity` which should be using `Data::NameAndTypeId`  (I think this maybe didn't exist when this PR was originally written?)

File: src/maidsafe/vault/mpid_manager/handler.h [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+#include "maidsafe/vault/chunk_store.h"
+#include "maidsafe/vault/mpid_manager/database.h"
+#include "maidsafe/vault/mpid_manager/messages.h"
+
+namespace maidsafe {
+
+namespace vault {
+
+typedef Identity MpidName;
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/handler.h]: We should prefer `using` to `typedef` now.

File: src/maidsafe/vault/mpid_manager/handler.h [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+using DbMessageQueryResult = boost::expected<MpidMessage, maidsafe_error>;
+using DbDataQueryResult = boost::expected<ImmutableData, maidsafe_error>;
+
+class MpidManagerHandler {
+ public:
+  MpidManagerHandler(const boost::filesystem::path vault_root_dir, DiskUsage max_disk_usage);
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/handler.h]: Should be const&?

File: src/maidsafe/vault/mpid_manager/handler.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  template <typename DataName>
+  void DeleteChunk(const DataName& data_name);
+
+  ChunkStore chunk_store_;
+  MpidManagerDatabase db_;
+};
+
+template <typename Data>
+Data MpidManagerHandler::GetChunk(const Identity& data_name) const {
+  typename Data::NameAndTypeId key(data_name, DataTypeId(0));
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/handler.h]: `DataTypeId(0)` is for `ImmutableData` only.

File: src/maidsafe/vault/mpid_manager/messages.cc [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+MpidMessageBase::MpidMessageBase(const MpidMessageBase& other)
+    : sender(other.sender), receiver(other.receiver), id(other.id), parent_id(other.parent_id),
+      signed_header(other.signed_header) {}
+
+MpidMessageBase::MpidMessageBase(MpidMessageBase&& other)
+    : sender(std::move(other.sender)), receiver(std::move(other.receiver)), id(std::move(other.id)),
+      parent_id(std::move(other.parent_id)), signed_header(std::move(other.signed_header)) {}
+
+MpidMessageBase& MpidMessageBase::operator=(MpidMessageBase other) {
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/messages.cc]: We don't use the copy-and-swap idiom any more.  The copy c'tor and copy assignment operator can be written `= default;`, but we need to hand-write the move ones for MSVC.

File: src/maidsafe/vault/mpid_manager/messages.cc [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+namespace maidsafe {
+
+namespace vault {
+
+// ================================= MpidMessageBase =============================================
+
+MpidMessageBase::MpidMessageBase()
+    : sender(), receiver(), id(), parent_id(), signed_header() {}
+
+MpidMessageBase::MpidMessageBase(const std::string& serialised_copy)
```
Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/messages.cc]: We shouldn't need this c'tor now we're using Cereal.

File: src/maidsafe/vault/vault.h [hunk 11]
```diff
@@ -39,18 +42,38 @@ namespace maidsafe {
...
+//  typename ParsedType::Name name;
+//  typename ParsedType::serialised_type contents;
+//  Parse(binary_input_stream, name, contents);
+//  return ParsedType(name, contents);
+// }
+
+// Helper function to parse Mpid messaging related messages : message, alert, etc.
+// FIXME this need discussion, adding it temporarily to progress
+template <typename ParsedType>
+ParsedType ParseMpidMessaging(const SerialisedData& serialised_data) {
```
Comment (Fraser999) [inline src/maidsafe/vault/vault.h]: Don't think we need this helper if we use the Cereal functions.  Just call `Parse` where you call `ParseMpidMessaging` and it should work.

File: src/maidsafe/vault/mpid_manager/handler.h (hunk 8 above)
Comment (maqi) [inline src/maidsafe/vault/mpid_manager/handler.h]: yes, the MpidMessage / MpidAlert will be stored as immutable chunk temporarily in MpidManager.
So DataTypeId(0) is to be used here.
Probably this function shall not be templated to reflect that.

File: src/maidsafe/vault/mpid_manager/database.h (hunk 3 above)
Comment (maqi) [inline src/maidsafe/vault/mpid_manager/database.h:80]: MpidManagerDatabase using multi_index as in-memory usage.
query to the multi_index incurs re-sorting of it internally, which considered as non-const.
the compiler won't let me put the functions as const

Comment (Fraser999) [inline src/maidsafe/vault/mpid_manager/database.h:80]: Oh - OK.  Thanks.

**Pull request merged**